Fix Matrix/Vector operator - and /, add tests

marci543 opened the pull request:
This PR fixes  element-wise `Matrix`/`Vector` operator `-` and `/` and adds tests.

It also closes #26.

michelp commented:
Thanks for catching these! I missed them when I did the upcasing of all the operator names.

Did you see Tim's email today about how he implements the overloaded - operator in MATLAB?  I like his argument and think you should do the same here.

michelp commented:
I'm going to drop Tim's email here so this isn't lost:

 can't answer for Python, but I do know how MATLAB would expect to work.  I put a lot of thought into this issue.  The basic rule is this:  If using an overloaded MATLAB operator, do the MATLAB thing; if using a GrB* function, do the GraphBLAS thing.

I would expect Python to act the same way, except that it can use a Python "with" context, which doesn't exist in MATLAB.

In MATLAB, C=A-B using an overloaded "-" operator does what MATLAB users would expect, so if A(i,j) is not present but B(i,j) is, then the result is C(i,j)=-B(i,j).   In particular, see the 'minus.m' function which supplies the C=A-B overloaded operator.  It's a simple one-liner:
https://github.com/DrTimothyAldenDavis/GraphBLAS/blob/stable/GraphBLAS/%40GrB/minus.m

That calls the uminus operator (in uminus.m):
https://github.com/DrTimothyAldenDavis/GraphBLAS/blob/stable/GraphBLAS/%40GrB/uminus.m
and then the plus.m operator:
https://github.com/DrTimothyAldenDavis/GraphBLAS/blob/stable/GraphBLAS/%40GrB/plus.m

The reason I do this is that all of the overloaded operators in my MATLAB interface use the conventional semiring and operators.  So C=A*B uses the PLUS_TIMES_[T] semiring, where T is the type of A.  C = A+B uses the GrB_PLUS_[T] operator, where T is the type of A.  (Another case where I have to query the matrix type, with the non-existent function GrB_Matrix_type, by the way... so I have to use my own GxB_Matrix_Type instead ...).  C=A.*B uses GrB_TIMES_[T], C=A-B uses GrB_MINUS_[T], and so on.

But C=GrB.eadd(A,'-',B) is a direct connection to the GraphBLAS GrB_eWiseAdd, and it would return C(i,j)=B(i,j) for this case.  Here, the user can pass in any operator, such as C = GrB.eadd ('max', A,B), for example.

But C = GrB.eadd ('max',A,B) is not the same as 'C=max(A,B)' in MATLAB.  So I do both.

The GrB_MAX_[T] operator is a case where eWiseAdd does something you might not expect, and it's not because max is not commutative.  The operator is applied just to the set intersection of the pattern of A and B, if you use GrB_eWiseAdd.  But in MATLAB, the built-in max(A,B) makes an assumption that the 'value of an entry not present' is zero.  I do what a MATLAB user would expect for max(A,B), however, which is "an entry not present has the value zero".  That is not the same as GrB.eadd('max',A,B).  Instead, I have to do this:
https://github.com/DrTimothyAldenDavis/GraphBLAS/blob/stable/GraphBLAS/%40GrB/max.m
(in particular, see lines 53 to 92 of the file max.m).  That does some initial work to expand scalars, if needed, then calls my sparse comparator helper function:
https://github.com/DrTimothyAldenDavis/GraphBLAS/blob/stable/GraphBLAS/%40GrB/private/gb_sparse_comparator.m
where in that function, I expand both A and B to the set union of A and B, and then to an eWiseAdd.
(I should actually to an eWiseMult at that point, since the pattern of A and B are identical, and my eWiseMult is now
faster than my eWiseAdd ... but that's another issue).

As a result, C=max(A,B) does what the MATLAB user would expect, for GrB matrices A and B (or if one of A or B are GrB and the other is a MATLAB matrix).  But C = GrB.eadd('max',A,B) does the exact GraphBLAS computation, which is different.

Since this might be a source of confusion, I wrote about in the MATLAB demo.
See https://github.com/DrTimothyAldenDavis/GraphBLAS/blob/bf6f8d77fa1c7d9aadcc8c042f884264d8521af1/GraphBLAS/demo/html/DellXPS13/v320/graphblas_demo.html#L410 (but it's easier to read in a browser):

marci543 commented:
Sorry for the late reply.
I rebased the branch and changed the behaviour of operator - accordingly. I also adjusted the tests.

The following questions come to my mind:
- What should we do with the explicit zeros in the result? Which one is better? 1-1=0 or 1-1=empty
- How should operator / behave?

marci543 commented:
Should we close this PR at this stage and discuss further questions in a separate issue?

michelp commented:
Hi @marci543 sorry for the delay!  I'll go ahead and merge this, i'm fine with a new issue to discuss it, might need to consult with Tim or his code.
